core/src/main/java/org/apache/iceberg/io/BaseDeltaWriter.java

```diff
@@ -0,0 +1,248 @@
+    }
+
+    if (posDeleteWriter != null && equalityDeleteWriter == null) {
+      // Only accept INSERT records and position deletion.
+      Preconditions.checkArgument(tableSchema == null);
+      Preconditions.checkArgument(equalityFieldIds == null);
+    }
+
+    if (equalityDeleteWriter != null) {
+      // Accept insert records, position deletion, equality deletions.
+      Preconditions.checkNotNull(posDeleteWriter,
+          "Position delete writer shouldn't be null when writing equality deletions.");
+      Preconditions.checkNotNull(tableSchema, "Iceberg table schema shouldn't be null");
+      Preconditions.checkNotNull(equalityFieldIds, "Equality field ids shouldn't be null");
+      Preconditions.checkNotNull(structLikeFun, "StructLike function shouldn't be null");
+
+      Schema deleteSchema = TypeUtil.select(tableSchema, Sets.newHashSet(equalityFieldIds));
+      this.insertedRowMap = StructLikeMap.create(deleteSchema.asStruct());
+      this.structLikeFun = structLikeFun;
+    } else {
+      this.insertedRowMap = null;
+      this.structLikeFun = null;
+    }
+
+    this.dataWriter = dataWriter;
+    this.equalityDeleteWriter = equalityDeleteWriter;
+    this.posDeleteWriter = posDeleteWriter;
+  }
+
+  @Override
+  public void writeRow(T row) throws IOException {
+    if (enableEqualityDelete()) {
+      FilePos filePos = FilePos.create(dataWriter.currentPath(), dataWriter.currentPos());
+      insertedRowMap.compute(structLikeFun.apply(row), (k, v) -> {
+        if (v == null) {
+          return Lists.newArrayList(filePos);
+        } else {
+          v.add(filePos);
+          return v;
+        }
+      });
+    }
+
+    dataWriter.write(row);
+  }
+
+  @Override
+  public void writeEqualityDelete(T equalityDelete) throws IOException {
+    if (!enableEqualityDelete()) {
+      throw new UnsupportedOperationException("Could not accept equality deletion.");
+    }
+
+    List<FilePos> existing = insertedRowMap.get(structLikeFun.apply(equalityDelete));
+
+    if (existing == null) {
+      // Delete the row which have been written by other completed delta writer.
+      equalityDeleteWriter.write(equalityDelete);
+    } else {
+      // Delete the rows which was written in current delta writer.
+      for (FilePos filePos : existing) {
+        posDeleteWriter.write(positionDelete.set(filePos.path(), filePos.pos(), null));
+      }
+    }
+  }
+
+  @Override
+  public void writePosDelete(CharSequence path, long offset, T row) throws IOException {
+    if (!enablePosDelete()) {
+      throw new UnsupportedOperationException("Could not accept position deletion.");
+    }
+
+    posDeleteWriter.write(positionDelete.set(path, offset, row));
+  }
+
+  @Override
+  public void abort() {
+    if (dataWriter != null) {
+      try {
+        dataWriter.abort();
+      } catch (IOException e) {
+        LOG.warn("Failed to abort the data writer {} because: ", dataWriter, e);
```

Comment:
I think abort should accept an optional exception, so that the suppressed exceptions can be added to it rather than just logged.

Reply:
I'm still thinking how to handle this gracefully.